drivers/input/input_cf1133.c

```diff
@@ -0,0 +1,358 @@
+#define X0_COORD_L          0x13
+#define Y0_COORD_L          0x14
+#define I2C_PROTOCOL        0x3E
+#define MAX_NUM_TOUCHES     0x3F
+#define DATA_0_HIGH         0x40
+#define DATA_0_LOW          0x41
+#define MISC_CONTROL        0xF1
+#define SMART_WAKE_UP_REG   0xF2
+#define CHIP_ID             0xF4
+#define PAGE_REG            0xff
+
+/* Constants  */
+#define SITRONIX_MAX_SUPPORTED_POINT 10U
+
+/* Mask  */
+#define ONE_D_SENSING_CONTROL_SHFT GENMASK(1, 1)
+#define ONE_D_SENSING_CONTROL_BMSK GENMASK(1, 0)
+#define I2C_PROTOCOL_BMSK          GENMASK(1, 0)
+
+/* Useful enum  */
+
+#define XY_COORD_H               0x0
+#define X_COORD_L                0x1
+#define Y_COORD_L                0x2
+#define PIXEL_DATA_LENGTH_B      0x3
+#define PIXEL_DATA_LENGTH_A      0x4
+#define SITRONIX_RESERVED_TYPE_0 0x0
+#define SITRONIX_A_TYPE          0x1
+#define SITRONIX_B_TYPE          0x2
+
+/** CF1133 configuration (DT). */
+struct cf1133_config {
+	/** I2C bus. */
+	struct i2c_dt_spec bus;
+	struct gpio_dt_spec reset_gpio;
+#ifdef CONFIG_INPUT_CF1133_INTERRUPT
+	/** Interrupt GPIO information. */
+	struct gpio_dt_spec int_gpio;
+#endif
+};
+
+/** CF1133 data. */
+struct cf1133_data {
+	/** Device pointer. */
+	const struct device *dev;
+	/** Work queue (for deferred read). */
+	struct k_work work;
+#ifdef CONFIG_INPUT_CF1133_INTERRUPT
+	/** Interrupt GPIO callback. */
+	struct gpio_callback int_gpio_cb;
+#else
+	/** Timer (polling mode). */
+	struct k_timer timer;
+#endif
+	/** Last pressed state. */
+	bool pressed_old;
+	bool pressed;
+
+	int resolution_x;
+	int resolution_y;
+	uint8_t max_touches;
+	uint8_t touch_protocol_type;
+	uint8_t pixel_length;
+	uint8_t chip_id;
+
+	uint8_t num_x;
+	uint8_t num_y;
+	uint8_t sensing_mode;
+};
+
+static int cf1133_get_chip_id(struct cf1133_data *data, const struct cf1133_config *config)
+{
+	int ret;
+	uint8_t buffer[3];
+
+	ret = i2c_burst_read_dt(&config->bus, CHIP_ID, buffer, sizeof(buffer));
+	if (ret < 0) {
+		LOG_ERR("Read burst failed: %d", ret);
+		return ret;
+	}
+	if (buffer[0] == 0) {
+		if (buffer[1] + buffer[2] > 32) {
+			data->chip_id = 2;
+		} else {
+			data->chip_id = 0;
+		}
+	} else {
+		data->chip_id = buffer[0];
+	}
+	data->num_x = buffer[1];
+	data->num_y = buffer[2];
+	LOG_DBG("Chip ID = %d, num_x = %d, num_y = %d ", data->chip_id, data->num_x, data->num_y);
+
+	return 0;
+}
+
+static int cf1133_get_protocol_type(struct cf1133_data *data, const struct cf1133_config *config)
+{
+	int ret = 0;
+	uint8_t buffer;
+
+	if (data->chip_id <= 3) {
+		ret = i2c_reg_read_byte_dt(&config->bus, I2C_PROTOCOL, &buffer);
+		if (ret < 0) {
+			LOG_ERR("read i2c protocol failed: %d", ret);
+			return ret;
+		}
+		data->touch_protocol_type = FIELD_GET(I2C_PROTOCOL_BMSK, buffer);
+		LOG_DBG("i2c protocol = %d ", data->touch_protocol_type);
+		data->sensing_mode =
+			(FIELD_GET((ONE_D_SENSING_CONTROL_BMSK << ONE_D_SENSING_CONTROL_SHFT),
+				   buffer)) >>
+			ONE_D_SENSING_CONTROL_SHFT;
```

Comment:
is this right? `FIELD_GET` is supposed to do the shifting for you

Reply:
Yes corrected